source-firestore/pull.go

```diff
@@ -54,261 +39,561 @@ func (driver) Pull(stream pc.Driver_PullServer) error {
+// fully synced, and this could potentially be many gigabytes of data.
+//
+// By emitting empty checkpoints periodically during the capture we unblock Flow
+// to persist our capture output instead of buffering everything, at the cost of
+// potentially duplicating documents in the event of a connector restart. I think
+// this is the best we can do, given the Firestore APIs and the constraint of not
+// buffering the entire dataset locally.
+const emptyCheckpoint string = `{}`
+
+func (c *capture) HandleDocument(ctx context.Context, resourcePath string, doc *firestore_pb.Document) error {
+	// Ignore document changes which occurred prior to the last read time of the collection.
+	var ctime = doc.CreateTime.AsTime()            // The time at which this document was first created
+	var mtime = doc.UpdateTime.AsTime()            // The time at which this document was last modified
+	var rtime, ok = c.State.ReadTime(resourcePath) // The latest read time for this resource path
+	if lvl := log.TraceLevel; log.IsLevelEnabled(lvl) {
+		log.WithFields(log.Fields{
+			"doc":   doc.Name,
+			"ctime": ctime.Format(time.RFC3339Nano),
+			"mtime": mtime.Format(time.RFC3339Nano),
+			"rtime": rtime.Format(time.RFC3339Nano),
+			"res":   resourcePath,
+		}).Log(lvl, "document change")
+	}
+	if !ok {
+		log.WithField("doc", doc.Name).Trace("ignoring document (resource not captured)")
+		return nil
+	}
+	if delta := mtime.Sub(rtime); delta < 0 {
+		log.WithField("doc", doc.Name).Trace("ignoring document (mtime < rtime)")
+		return nil
+	}
+
+	// Convert the document into a JSON-serializable map of fields
+	var fields = make(map[string]interface{})
+	for id, val := range doc.Fields {
+		var tval, err = translateValue(val)
 		if err != nil {
-			return err
-		}
-		c.Output <- emitRequest{
-			requestType: EncodeDocument,
-			binding:     uint32(index),
-			arena:       docJson,
+			return fmt.Errorf("error translating value: %w", err)
 		}
+		fields[id] = tval
+	}
+	fields[metaProperty] = &documentMetadata{
+		Path:       doc.Name,
+		CreateTime: &ctime,
+		UpdateTime: &mtime,
 	}
 
-	// If we are not tailing, we close the channel after the full scan
-	// to allow the encoding routine to finish so the connector can exit
-	if !c.Tail {
+	if bindingIndex, ok := c.State.BindingIndex(resourcePath); !ok {
+		// Listen streams can be a bit over-broad. For instance if there are
+		// collections 'users/*/docs' and 'groups/*/docs' in the database, but
+		// only 'users/*/docs' is captured, we need to ignore any documents
+		// from paths like 'groups/*/docs' which don't map to any binding.
 		return nil
+	} else if docJSON, err := json.Marshal(fields); err != nil {
+		return fmt.Errorf("error serializing document %q: %w", doc.Name, err)
+	} else if err := c.Output.Documents(bindingIndex, docJSON); err != nil {
+		return err
+	} else if err := c.Output.Checkpoint(json.RawMessage(emptyCheckpoint), true); err != nil {
```

Comment:
Does this ever end up "voiding" previous non-empty checkpoints? Might be part of the necessary design, just wanting to make sure I understand correctly.

Reply:
No, the `true` argument there says "this is a patch update, to be merged according to https://www.rfc-editor.org/rfc/rfc7396 semantics", and the empty checkpoint is the identity value under merge semantics. So in effect this just says "hey Flow here's a checkpoint so you can finish the transaction whenever it's convenient" without actually modifying the state contents.

Reply:
Ah thanks! Makes sense.